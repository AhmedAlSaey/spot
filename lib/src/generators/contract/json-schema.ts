import assertNever from "assert-never";
import { Type } from "../../models";
import compact = require("lodash/compact");

export function noVoidJsonSchema(
  flavour: Flavour,
  type: Type,
  errorMessage: string
): JsonSchemaType {
  const jsonSchemaType = jsonSchema(flavour, type);
  if (!jsonSchemaType) {
    throw new Error(errorMessage);
  }
  return jsonSchemaType;
}

export function voidToNullJsonSchema(
  flavour: Flavour,
  type: Type
): JsonSchemaType {
  const jsonSchemaType = jsonSchema(flavour, type);
  if (!jsonSchemaType) {
    return {
      type: "null"
    };
  }
  return jsonSchemaType;
}

export type Flavour = "json-schema-draft-7" | "openapi-3";

export const FLAVOURS: Flavour[] = ["json-schema-draft-7", "openapi-3"];

export function jsonSchema(
  flavour: Flavour,
  type: Type
): JsonSchemaType | null {
  switch (type.kind) {
    case "void":
      return null;
    case "null":
      return {
        type: "null"
      };
    case "boolean":
      return {
        type: "boolean"
      };
    case "boolean-constant":
      return {
        type: "boolean",
        const: type.value
      };
    case "string":
      return {
        type: "string"
      };
    case "string-constant":
      return {
        type: "string",
        const: type.value
      };
    case "number":
      return {
        type: "number"
      };
    case "integer-constant":
      return {
        type: "integer",
        const: type.value
      };
    case "object":
      return Object.entries(type.properties).reduce(
        (acc, [name, type]) => {
          if (type.kind === "optional") {
            type = type.optional;
          } else {
            acc.required.push(name);
          }
          const jsonSchemaType = jsonSchema(flavour, type);
          if (jsonSchemaType) {
            acc.properties[name] = jsonSchemaType;
          }
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: []
        } as JsonSchemaObject & { required: string[] }
      );
    case "array":
      const itemsType = jsonSchema(flavour, type.elements);
      if (!itemsType) {
        throw new Error(`Unsupported void array`);
      }
      return {
        type: "array",
        items: itemsType
      };
    case "optional":
      throw new Error(`Unsupported top-level optional type`);
    case "union":
      const types = type.types.map(t => jsonSchema(flavour, t));
      const withoutNullTypes = compact(types);
      if (withoutNullTypes.length !== types.length) {
        throw new Error(`Unsupported void type in union`);
      }
      return {
        oneOf: withoutNullTypes
      };
    case "type-reference":
      switch (flavour) {
        case "json-schema-draft-7":
          return {
            $ref: `#/definitions/${type.typeName}`
          };
        case "openapi-3":
          return {
            $ref: `#/components/schemas/${type.typeName}`
          };
        default:
          throw assertNever(flavour);
      }
    default:
      throw assertNever(type);
  }
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
