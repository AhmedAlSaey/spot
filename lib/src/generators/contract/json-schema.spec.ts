import {
  arrayType,
  BOOLEAN,
  booleanConstant,
  integerConstant,
  NULL,
  NUMBER,
  objectType,
  optionalType,
  STRING,
  stringConstant,
  typeReference,
  unionType,
  VOID
} from "../../models";
import { FLAVOURS, jsonSchema } from "./json-schema";

describe("JSON Schema generator", () => {
  describe("generates type validator", () => {
    for (const flavour of FLAVOURS) {
      test(`${flavour} - void`, () => {
        expect(jsonSchema(flavour, VOID)).toMatchSnapshot();
      });

      test(`${flavour} - null`, () => {
        expect(jsonSchema(flavour, NULL)).toMatchSnapshot();
      });

      test(`${flavour} - boolean`, () => {
        expect(jsonSchema(flavour, BOOLEAN)).toMatchSnapshot();
      });

      test(`${flavour} - boolean constant`, () => {
        expect(jsonSchema(flavour, booleanConstant(true))).toMatchSnapshot();
        expect(jsonSchema(flavour, booleanConstant(false))).toMatchSnapshot();
      });

      test(`${flavour} - string`, () => {
        expect(jsonSchema(flavour, STRING)).toMatchSnapshot();
      });

      test(`${flavour} - string constant`, () => {
        expect(
          jsonSchema(flavour, stringConstant("some constant"))
        ).toMatchSnapshot();
      });

      test(`${flavour} - number`, () => {
        expect(jsonSchema(flavour, NUMBER)).toMatchSnapshot();
      });

      test(`${flavour} - integer constant`, () => {
        expect(jsonSchema(flavour, integerConstant(0))).toMatchSnapshot();
        expect(jsonSchema(flavour, integerConstant(123))).toMatchSnapshot();
        expect(jsonSchema(flavour, integerConstant(-1000))).toMatchSnapshot();
      });

      test(`${flavour} - object`, () => {
        expect(jsonSchema(flavour, objectType({}))).toMatchSnapshot();
        expect(
          jsonSchema(
            flavour,
            objectType({
              singleField: NUMBER
            })
          )
        ).toMatchSnapshot();
        expect(
          jsonSchema(
            flavour,
            objectType({
              field1: NUMBER,
              field2: STRING,
              field3: optionalType(BOOLEAN)
            })
          )
        ).toMatchSnapshot();
      });

      test(`${flavour} - array`, () => {
        expect(jsonSchema(flavour, arrayType(STRING))).toMatchSnapshot();
      });

      test(`${flavour} - optional`, () => {
        expect(() => jsonSchema(flavour, optionalType(STRING))).toThrowError(
          "Unsupported top-level optional type"
        );
      });

      test(`${flavour} - union`, () => {
        expect(
          jsonSchema(flavour, unionType(STRING, NUMBER, BOOLEAN))
        ).toMatchSnapshot();
      });

      test(`${flavour} - type reference`, () => {
        expect(
          jsonSchema(flavour, typeReference("OtherType"))
        ).toMatchSnapshot();
      });
    }
  });
});
