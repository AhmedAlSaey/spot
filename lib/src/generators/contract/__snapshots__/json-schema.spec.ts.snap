// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Schema generator generates type validator json-schema-draft-7 - array 1`] = `
Object {
  "items": Object {
    "type": "string",
  },
  "type": "array",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - boolean 1`] = `
Object {
  "type": "boolean",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - boolean constant 1`] = `
Object {
  "const": true,
  "type": "boolean",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - boolean constant 2`] = `
Object {
  "const": false,
  "type": "boolean",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - integer constant 1`] = `
Object {
  "const": 0,
  "type": "integer",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - integer constant 2`] = `
Object {
  "const": 123,
  "type": "integer",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - integer constant 3`] = `
Object {
  "const": -1000,
  "type": "integer",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - null 1`] = `
Object {
  "type": "null",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - number 1`] = `
Object {
  "type": "number",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - object 1`] = `
Object {
  "properties": Object {},
  "required": Array [],
  "type": "object",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - object 2`] = `
Object {
  "properties": Object {
    "singleField": Object {
      "type": "number",
    },
  },
  "required": Array [
    "singleField",
  ],
  "type": "object",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - object 3`] = `
Object {
  "properties": Object {
    "field1": Object {
      "type": "number",
    },
    "field2": Object {
      "type": "string",
    },
    "field3": Object {
      "type": "boolean",
    },
  },
  "required": Array [
    "field1",
    "field2",
  ],
  "type": "object",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - string 1`] = `
Object {
  "type": "string",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - string constant 1`] = `
Object {
  "const": "some constant",
  "type": "string",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - type reference 1`] = `
Object {
  "$ref": "#/definitions/OtherType",
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - union 1`] = `
Object {
  "oneOf": Array [
    Object {
      "type": "string",
    },
    Object {
      "type": "number",
    },
    Object {
      "type": "boolean",
    },
  ],
}
`;

exports[`JSON Schema generator generates type validator json-schema-draft-7 - void 1`] = `null`;

exports[`JSON Schema generator generates type validator openapi-3 - array 1`] = `
Object {
  "items": Object {
    "type": "string",
  },
  "type": "array",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - boolean 1`] = `
Object {
  "type": "boolean",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - boolean constant 1`] = `
Object {
  "const": true,
  "type": "boolean",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - boolean constant 2`] = `
Object {
  "const": false,
  "type": "boolean",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - integer constant 1`] = `
Object {
  "const": 0,
  "type": "integer",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - integer constant 2`] = `
Object {
  "const": 123,
  "type": "integer",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - integer constant 3`] = `
Object {
  "const": -1000,
  "type": "integer",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - null 1`] = `
Object {
  "type": "null",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - number 1`] = `
Object {
  "type": "number",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - object 1`] = `
Object {
  "properties": Object {},
  "required": Array [],
  "type": "object",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - object 2`] = `
Object {
  "properties": Object {
    "singleField": Object {
      "type": "number",
    },
  },
  "required": Array [
    "singleField",
  ],
  "type": "object",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - object 3`] = `
Object {
  "properties": Object {
    "field1": Object {
      "type": "number",
    },
    "field2": Object {
      "type": "string",
    },
    "field3": Object {
      "type": "boolean",
    },
  },
  "required": Array [
    "field1",
    "field2",
  ],
  "type": "object",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - string 1`] = `
Object {
  "type": "string",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - string constant 1`] = `
Object {
  "const": "some constant",
  "type": "string",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - type reference 1`] = `
Object {
  "$ref": "#/components/schemas/OtherType",
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - union 1`] = `
Object {
  "oneOf": Array [
    Object {
      "type": "string",
    },
    Object {
      "type": "number",
    },
    Object {
      "type": "boolean",
    },
  ],
}
`;

exports[`JSON Schema generator generates type validator openapi-3 - void 1`] = `null`;
